Added cache data collector and profiler page

Nyholm opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #19297 
| License       | MIT
| Doc PR        | n/a

Adding a first version of cache profiler page. Most things are taken from PHP-cache. 
FYI: @aequasi


### What is included?

A collector, recording adapter and a profiler page. 

![screen shot 2016-12-27 at 16 07 35](https://cloud.githubusercontent.com/assets/1275206/21502325/4bee2ed4-cc4f-11e6-89fc-37ed16aca864.png)
![screen shot 2016-12-27 at 16 07 45](https://cloud.githubusercontent.com/assets/1275206/21502326/4bee9450-cc4f-11e6-904d-527b7b0ce85b.png)

### What is not included?

* A good logo 
* Nice design on the profiler page

This PR aims to pass as the minimum requirement for a cache page. Im happy to do fancy extra features but those should be a different PR.

Nyholm committed:
Added cache data collector and profiler page

Nyholm commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -645,7 +646,9 @@ private function registerTemplatingConfiguration(array $config, ContainerBuilder
-            $loaders = array_map(function ($loader) { return new Reference($loader); }, $config['loaders']);
+            $loaders = array_map(function ($loader) {
+                return new Reference($loader);
+            }, $config['loaders']);
```
This change was from fabbot.io

nicolas-grekas replied:
you should revert them, fabbot is broken these days

Nyholm replied:
Okey, thank you

Nyholm commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -674,7 +677,9 @@ private function registerTemplatingConfiguration(array $config, ContainerBuilder
 
         $container->setParameter('templating.engines', $config['engines']);
-        $engines = array_map(function ($engine) { return new Reference('templating.engine.'.$engine); }, $config['engines']);
+        $engines = array_map(function ($engine) {
```
This change was from fabbot.io

javiereguiluz commented:
@Nyholm thanks for this PR. Don't worry about the logo because the SensioLabs designer will take care of that. And same thing about the design, don't worry because we can iterate it later. Let's focus on the contents and the information provided. Thanks!

Nyholm commented:
That is great. Then Im done with this PR =)

Koc commented:
It would be very useful collect this information also from production env and analyze it later. Can anybody idea how to make it possible without enabling whole profiler?

Nyholm commented:
>It would be very useful collect this information also from production env and analyze it later. Can anybody idea how to make it possible without enabling whole profiler?

With the current implementation you should only load `cache_debug.xml` and then make sure to store the data from the collector somewhere on `kernel.terminate`

Renrhaf commented on src/Symfony/Component/Cache/Adapter/RecordingAdapter.php:
```diff
@@ -0,0 +1,194 @@
+     *
+     * @return string
+     */
+    private function getValueRepresentation($value)
```
I'm not sure, but can't we use the VarDumper component here to print the value ?

Nyholm replied:
Oh. That is a good idea. I'll see what I can do. Thanks.

nicolas-grekas replied:
the representation should be handled by the data collector, using the cloneVar method of the base DataCollector

Nyholm replied:
I cannot use the `DataCollector::cloneVar` in this context. The `RecordingAdapter` is not a data collector. However, I use the `cloneVar` once the data is moved to the `CacheDataCollector`

Nyholm replied:
@Renrhaf I've updated the PR using your suggestion. Thank you!

Renrhaf replied:
@Nyholm seems good to me ;) thank you !

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/RecordingAdapter.php:
```diff
@@ -0,0 +1,194 @@
+     *
+     * @return object
+     */
+    private function timeCall($name, array $arguments = array())
```
what about using a stopwatch instead?

Nyholm replied:
Sure!

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/RecordingAdapter.php:
```diff
@@ -0,0 +1,194 @@
+     *
+     * @param $call
+     */
+    private function addCall($call)
```
this function wrapper for a oneliner could be removed

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/RecordingAdapterInterface.php:
```diff
@@ -0,0 +1,24 @@
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+interface RecordingAdapterInterface
```
I'd remove this interface entirely: we don't want a ontract here, we need only a behavior (an implem)

nicolas-grekas commented on src/Symfony/Component/Cache/DataCollector/CacheDataCollector.php:
```diff
@@ -0,0 +1,159 @@
+            foreach ($calls as $call) {
+                $statistics[$name]['calls'] += 1;
+                $statistics[$name]['time'] += $call->time;
+                if ($call->name === 'getItem') {
```
what about getItems?

Nyholm replied:
Thank you. If one do getItems('a', 'b', 'c'). I will count that as 1 call and 3 reads.

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/RecordingAdapter.php:
```diff
@@ -0,0 +1,202 @@
+        $result = $call->result;
+        $hits = 0;
+        $items = array();
+        foreach ($result as $item) {
```
In the case of a generator, I see two issues here:
- this is consuming the generator. Since these are not rewindable, `return $result;` returns a non usable generator.
- this is couting hits, but maybe the app didn't consume the generator on its side.

This means I think we should have two code paths here: one for array (doing the same as the current code),
and one for Traversable, which should wrap this in a generator-closure, and do things lazily.

Nyholm replied:
I've wrapped the result in a generator closure as you suggested. I fail to understand we need a code path for arrays. I know that getItems will return something that you can do foreach with. Why can't I just always wrap it in a generator-closure?

nicolas-grekas replied:
no strong reason, good also via a generator

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/RecordingAdapter.php:
```diff
@@ -0,0 +1,202 @@
+ * @author Aaron Scherer <aequasi@gmail.com>
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+final class RecordingAdapter implements AdapterInterface
```
Oh, I guess this should be named TraceableAdapter fr consistency with the code base (eg TraceableUrlMatcher, etc.)

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/TraceableAdapter.php:
```diff
@@ -0,0 +1,210 @@
+     *
+     * @return string
+     */
+    private function getValueRepresentation($value)
```
I think we don't need this logic, let's just use `$value` directly.
The clone may fail (implem can throw in `__clone`), and is fragile (not recursive).
and for other types of data, we don't care (resources cannot come up from a cache pool)
removing also will make the layer slimmer, which is also a nice target

Nyholm replied:
Will do. Thank you.

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CacheCollectorPass.php:
```diff
@@ -0,0 +1,53 @@
+<?php
+
+/*
+ * This file is part of php-cache\cache-bundle package.
```
this is wrong

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CacheCollectorPass.php:
```diff
@@ -0,0 +1,53 @@
+/*
+ * This file is part of php-cache\cache-bundle package.
+ *
+ * (c) 2015-2015 Aaron Scherer <aequasi@gmail.com>, Tobias Nyholm <tobias.nyholm@gmail.com>
```
Symfony code must use the Symfony copyright

Nyholm replied:
Sorry, Copy/paste issue.. An my IDE does not show file headers. I will fix these

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/cache.html.twig:
```diff
@@ -0,0 +1,120 @@
+{% block head %}
+    {{ parent() }}
+    <style>
+        .color-red {
```
would be better to include necessary styles in the CSS, to be usable for all collectors

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/cache.html.twig:
```diff
@@ -0,0 +1,120 @@
+        {{ include('@WebProfiler/Icon/cache.svg') }}
+    </span>
+    <strong>Cache</strong>
+    <span class="count">
```
I don't think we should have the time here anymore, for consistency with other panels

Nyholm replied:
Do you want me to remove it?

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/cache.html.twig:
```diff
@@ -0,0 +1,120 @@
+
+{% block panel %}
+    <h2>Cache</h2>
+    {% for name, calls in collector.calls %}
```
It would be great to have the aggregated stats first as big numbers (as done in other panels), before having the detail for each pool

Nyholm replied:
Done!

stof replied:
no it is not done

Nyholm replied:
Sorry, I misunderstood you. I put the aggregated stats for each adapter as **big numbers**. 
<img width="334" alt="screen shot 2017-01-02 at 18 46 11" src="https://cloud.githubusercontent.com/assets/1275206/21593831/d787e3c4-d11b-11e6-8636-998abf1abb85.png">

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/TraceableAdapter.php:
```diff
@@ -0,0 +1,207 @@
+namespace Symfony\Component\Cache\Adapter;
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
```
not used anymore (mybad)

stof commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CacheCollectorPass.php:
```diff
@@ -0,0 +1,53 @@
+            $container->register($id.'.recorder', TraceableAdapter::class)
+                ->setDecoratedService($id)
+                ->addArgument(new Reference($id.'.recorder.inner'))
+                ->addArgument(new Reference('debug.stopwatch'))
```
The implementation does not use Stopwatch

Nyholm commented:
![screen shot 2016-12-28 at 20 39 35](https://cloud.githubusercontent.com/assets/1275206/21530003/ef2a6fec-cd3d-11e6-8f7d-de6228e9fde8.png)
![screen shot 2016-12-28 at 20 39 25](https://cloud.githubusercontent.com/assets/1275206/21530004/ef2add56-cd3d-11e6-8ee8-6653707774ef.png)

xabbuh commented:
Just by looking at the screenshot I was not able to easily recognise what "ratio" refers to. Does it mean hits/reads?

Nyholm commented:
Sorry about that. Yes, the ratio is the hits/reads ratio. See new screenshot: 

<img width="1094" alt="screen shot 2016-12-28 at 20 59 03" src="https://cloud.githubusercontent.com/assets/1275206/21530377/b1f830de-cd40-11e6-931e-1381c2b33ec3.png">



------

FYI: I created a [separate PR](https://github.com/symfony/symfony/issues/21082) with the adapters. I added some tests and fixed some bugs in it. I will rebase this PR when/if #21082 gets merged.

xabbuh commented:
I think I would change the label from "ratio" to "hits/reads" then.

xabbuh commented:
And what about providing aggregated data across all pools?

Nyholm commented:
I will. 

> And what about providing aggregated data across all pools?

That is only in the toolbar atm.

ro0NL commented:
Nice :)

What about a tab per pool? That should organize things right?

The table with calls looks rather technical, so maybe we can improve that UX-wise.

- Try to keep simple `key | value | time` headers (only save calls seem to have extra metadata)
- Differ with colors between gets, deletes or saves (an in-row label or so)

:+1:

Nyholm committed:
Make sure we count hits/reads from hasItem

Nyholm commented:
Thank you @ro0NL. I agree with your comments that this could be improved and I really like your suggestions. I think this is "good enough" and I will let other PRs improve this. Does it sound reasonable?

I've rebased this PR on master.

Nyholm commented on src/Symfony/Component/Cache/DataCollector/CacheDataCollector.php:
```diff
@@ -0,0 +1,177 @@
+                    $statistics[$name]['misses'] += $count - $call->misses;
+                } elseif ($call->name === 'hasItem') {
+                    $statistics[$name]['reads'] += 1;
+                    if ($call->result->getRawData()[0][0] === false) {
```
This looks a bit ugly. 

`$call->result` is a `Symfony\Component\VarDumper\Cloner\Data`. I know that `hasItem` will always return a boolean thanks to PSR-6. So I believe this is safe.

nicolas-grekas replied:
No better way for now, I may work on something better soon but this is the way to go for now.

ro0NL commented:
> Does it sound reasonable?

Totally :) just aiming to get this shipped as pretty as possible with the first tag release :)

Nyholm commented:
The deps-low test is failing because Framework bundle fetches symfony/cache 3.2 (not dev-master). 

Im not sure how to fix that or if I should fix that.

javiereguiluz commented:
@Nyholm the designer of SensioLabs has just send me icon for the cache. It uses the same conceptualization as yours (because it's great) but it tweaks the design a bit to match the style of the other icons:


```xml
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path fill="#AAA" d="M2.26 6.09l9.06-4.67a1.49 1.49 0 0 1 1.37 0l9.06 4.67a1.49 1.49 0 0 1 0 2.65l-9.06 4.67a1.49 1.49 0 0 1-1.37 0L2.26 8.74a1.49 1.49 0 0 1 0-2.65zM20.55 11L12 15.39 3.45 11a1.36 1.36 0 0 0-1.25 2.42l9.17 4.73a1.36 1.36 0 0 0 1.25 0l9.17-4.73A1.36 1.36 0 0 0 20.55 11zm0 4.47L12 19.86l-8.55-4.41a1.36 1.36 0 0 0-1.25 2.42l9.17 4.73a1.36 1.36 0 0 0 1.25 0l9.17-4.73a1.36 1.36 0 0 0-1.25-2.42z"/></svg>
```

Nyholm committed:
Update logo

Nyholm commented:
I've updated the logos:

Before: 
![screen shot 2016-12-29 at 15 52 43](https://cloud.githubusercontent.com/assets/1275206/21546467/39aad970-cddf-11e6-9f10-e8bfeb0c2e2d.png)

After:
![screen shot 2016-12-29 at 15 52 33](https://cloud.githubusercontent.com/assets/1275206/21546460/34aaac2a-cddf-11e6-8a8b-11707b22f8b4.png)

FYI, It was @mablae that created the original logo for PHP-cache. Thank you.

xabbuh commented:
@Nyholm You can change the version constraint for `symfony/cache` to `~3.3`.

Nyholm commented:
Thank you @xabbuh It worked.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CacheCollectorPass.php:
```diff
@@ -0,0 +1,52 @@
+        $collectorDefinition = $container->getDefinition('data_collector.cache');
+        $serviceIds = $container->findTaggedServiceIds('cache.pool');
+
+        foreach (array_keys($serviceIds) as $id) {
```
`foreach ($container->findTaggedServiceIds('cache.pool') as $id => $service) {`, this will allow you to remove the call to getDefinition just below

aequasi replied:
Relic from old code i believe. Good to change

Nyholm replied:
```php
foreach ($container->findTaggedServiceIds('cache.pool') as $id => $service) {
  // $service is an array of tags.
  var_dump($service);
}

array(2) {
  [0] =>
  array(0) {
  }
  [1] =>
  array(2) {
    'clearer' =>
    string(21) "cache.default_clearer"
    'unlazy' =>
    bool(true)
  }
}

```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CacheCollectorPass.php:
```diff
@@ -0,0 +1,52 @@
+                ->setPublic(false);
+
+            // Tell the collector to add the new instance
+            $collectorDefinition->addMethodCall('addInstance', array($id, new Reference($id)));
```
what abaut reporting the "clearer", "provider", "namespace" and/or "default_lifetime" attributes when available?

nicolas-grekas replied:
Note also that this will prevent private services from being removed at compile time.
We have two options here: keep it this way so that even non-used private service could be seen in the panel,
or register this pass just before CachePoolClearerPass.

Nyholm replied:
> what about reporting the "clearer", "provider", "namespace" and/or "default_lifetime" attributes when available?

Sure. But can that be left out of this PR as a future improvement?

> or register this pass just before CachePoolClearerPass.

What do you prefer? Should I change the order of the passes

nicolas-grekas commented on src/Symfony/Component/Cache/DataCollector/CacheDataCollector.php:
```diff
@@ -0,0 +1,177 @@
+            foreach ($calls as $call) {
+                $statistics[$name]['calls'] += 1;
+                $statistics[$name]['time'] += $call->end - $call->start;
+                if ($call->name === 'getItem') {
```
yoda style everywhere please

stof commented:
> > And what about providing aggregated data across all pools?
> 
> That is only in the toolbar atm.

That's unfortunate, as it means that the main info displayed in the toolbar is not available anymore when looking at details.
Btw, I made the same comment earlier about that (and you said it was done while it is not

Nyholm committed:
yoda style

Nyholm committed:
Added large summary numbers

Nyholm committed:
rewite to show attributes

Nyholm commented:
Ping @stof 

![screen shot 2017-01-03 at 10 42 03](https://cloud.githubusercontent.com/assets/1275206/21604096/94eab0b0-d1a1-11e6-9d8f-e95e07057b16.png)

Nyholm committed:
Changed order

xabbuh commented:
@Nyholm Any reason not to aggregate missed and deletes? I would also change the order to be consistent with how it is for the cache pools.

Nyholm committed:
Added deletes

Nyholm commented:
I updated the order and added deletes. Thank you for the quick feedback. 
![screen shot 2017-01-03 at 10 52 16](https://cloud.githubusercontent.com/assets/1275206/21604307/bded3112-d1a2-11e6-9e4c-8b5595ed124a.png)

rvanlaak commented:
We chain a `apcu` pool with a `filesystem` pool, which means we have three pools in the profiler overview. It would be nice (for a future PR) to think about some chain-specific aggregates, or maybe allow to disable profiler data collection per pool?

From DX perspective it also would be great to have some kind of (javascript?) filtering in the panel, as that would allow you to instantly see the behavior path of a specific cache item.

Nyholm commented:
Is there anything else I can do to help getting this PR merged?

fabpot commented:
@Nyholm If you can rebase on current master, I can merge it :)

nicolas-grekas commented:
:+1: from me
we should open an issue with the remaining ideas to not forget about them
for this next issue : about the compiler pass, ideally, we could keep track of removed private pools so that the panel show them eg grayed.

Nyholm committed:
Fixed rebase

Nyholm commented:
Thank you. The PR is rebased and tests are all green.

fabpot commented:
Thank you @Nyholm.

Nyholm commented:
Thank you for merging

Renrhaf commented:
Many thanks @Nyholm :)
